### PR TITLE
Build Tooling: Update development tools to work with PostCSS 8

### LINKED ISSUE #27028: Update @wordpress/scripts to use PostCSS 8
Many PostCSS plugins are now requiring PostCSS 8, and so resources using `@wordpress/scripts` are failling to `npm run build` with latest module versions, throwing errors like:

```Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ./node_modules/postcss-loader/src/index.js):
Error: PostCSS plugin postcss-import requires PostCSS 8.
```

I think it's enough to bump the version of `postcss-loader` in `@wordpress/scripts` to  >= 4.0.3.

### PR DESCRIPTION
**Pull request by ocean90**
## Description
Fixes #27028.

This should allow to use @wordpress/scripts with PostCSS 8 but doesn't actually require it since it doesn't have it as a dependency at all.

See also https://github.com/webpack-contrib/postcss-loader/blob/v4.1.0/CHANGELOG.md

## How has this been tested?

`npm run dev` and `npm run build` still work properly,

Create Block still build correctly. Testing is a bit cumbersome but it follows what happens on CI (https://github.com/WordPress/gutenberg/blob/master/bin/test-create-block.sh):

1. `npx wp-create-block esnext-example --no-wp-scripts`
2. `cd esnext-example`
3. `../node_modules/.bin/wp-scripts start`
4. Make sure there are no errors and files are built in `build` folder.
5. `../node_modules/.bin/wp-scripts build`
6. Make sure there are no errors and files are built in `build` folder.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by ocean90** [comment]
From the changelog:

> `postcss` was moved to `peerDependencies`, you need to install `postcss`

So have to require this now by ourself which means we could switch to PostCSS 8. This would also mean we have to update [postcss-themes](https://github.com/WordPress/gutenberg/tree/master/packages/postcss-themes).

File: package-lock.json [hunk 1]
```diff
@@ -13050,6 +13050,94 @@
...
+					"resolved": "https://registry.npmjs.org/loader-utils/-/loader-utils-2.0.0.tgz",
+					"integrity": "sha512-rP4F0h2RaWSvPEkD7BLDFQnvSf+nK+wr3ESUjNTyAGobqrijmW92zc+SO6d4p4B1wh7+B/Jg1mkQe5NYUEHtHQ==",
+					"dev": true,
+					"requires": {
+						"big.js": "^5.2.2",
+						"emojis-list": "^3.0.0",
+						"json5": "^2.1.2"
+					}
+				},
+				"postcss-loader": {
```
Comment (gziolo) [inline package-lock.json]: I see `postcss-loader` loader wasn't hoisted to the top level `node_modules`. I wish there was a tool to automate updating npm dependencies in monorepo. I was using the following workaround recently:

1. Update the version of the dependnecy in all local packages.
2. Add the same dependency in the top-level `package.json` with the exact version.
3. Run `npm install`.
4. Remove the dependency from the top-level `package.json`.
5. Run `npm install` again.

It doesn't always work but increases chances it properly regenerates the lock file.

Maybe we need a script that goes through the lock file and removes all occurrences of the dependency with the version we want to remove as well. It'd be the most useful when doing upgrades for Babel, Jest or ESLint that have several packages that depend on each other. 

cc @jsnajdr

**Comment by gziolo** [comment]
> So have to require this now by ourself which means we could switch to PostCSS 8. This would also mean we have to update postcss-themes.

Yes, it looks like it's necessary. It also indirectly satisfies the need for having `postcss` as the dependency. I'm looking forward to the day when npm 7 becomes the default client for npm so we don't have to worry about installing peer dependencies in our packages (https://github.com/npm/rfcs/blob/latest/implemented/0025-install-peer-deps.md).

**Comment by gziolo** [comment]
With 41fcfd8 I tried to upgrade `postcss` to v8, but I also updated several other packages used by webpack which made it evident that they don't quite work with webpack 5...

I will try again with fewer changes applied and a separate effort to fix `wp-scripts` with webpack 5.

**Comment by gziolo** [comment]
Now that revert for webpack 5 upgrade landed with https://github.com/WordPress/gutenberg/pull/27974, I updated this PR to work with weback 4 and PostCSS 8. In my testing, it works correctly with Gutenberg, but also with Create Block.

**Review by gziolo — APPROVED** [review]
@ocean90, it should be good to go, but I would appreciate sanity check :)

**Comment by ocean90** [comment]
@gziolo Did you see my comment in https://github.com/WordPress/gutenberg/issues/27028#issuecomment-748456056?

**Comment by gziolo** [comment]
> @gziolo Did you see my comment in [#27028 (comment)](https://github.com/WordPress/gutenberg/issues/27028#issuecomment-748456056)?

I missed it :(

It looks like all those issues are still present:

> To make sure we can use PostCSS 8 we have to wait a bit I think
> 
> * [`@storybook/core`](https://github.com/storybookjs/storybook/tree/master/lib/core) requires `postcss-loader@3`, see also https://github.com/storybookjs/storybook/issues/12668.
> * [`@wordpress/postcss-themes`](https://github.com/WordPress/gutenberg/tree/master/packages/postcss-themes) has a hard requirement on `postcss@7`. This should probably be changed to `peerDependencies`, see also https://evilmartians.com/chronicles/postcss-8-plugin-migration.
> * [postcss-custom-properties](https://github.com/postcss/postcss-custom-properties/blob/master/package.json), used by `@wordpress/postcss-themes` has also a hard requirement on `postcss@7`. Open pull request at https://github.com/postcss/postcss-custom-properties/pull/229.

**Comment by gziolo** [comment]
I rebase this branch with the latest changes in `trunk`. I updated all previously updated dependencies to their latest versions. In addition to that I applied two remaining changes from the list:
- `@wordpress/postcss-themes` has no longer a hard requirement on `postcss@7`.  It defines `postcss@8` as a peer dependency instead.
- `postcss-custom-properties` used by `@wordpress/postcss-themes` was updated to the latest version that works with `postcss@8`.

We are only waiting for the next version of Storybook `6.2.x` that will work with `postcss@8`. The related issue is already resolved: https://github.com/storybookjs/storybook/issues/12668. There are doing RC releases at the moment.

**Comment by gziolo** [comment]
I executed `npx sb upgrade --prerelease` and everything works now. I think this PR is ready for the final review. We might still want to wait a few days/weeks before Storybook releases the stable version of `v6.2.x`

There is only 5 blocking items to finish to see the stable version of Storybook:
https://github.com/storybookjs/storybook/milestone/56

**Comment by gziolo** [comment]
Updated Storybook to the RC8 in fa54c88. There are 3 issues left in the milestone that block its stable release that don't impact Gutenberg.

**Review by gwwar — APPROVED** [review]
Thanks for looking into this. This smoke tests okay for me 👍  Folks might want to double check that package-lock.json is generated correctly, since the changes look a little large.

I tested this with:
```bash
npm run distclean
npm run install
npm run build
npm run dev
npm run storybook:build
npm run storybook:dev
npm run test:create-block
```
Output of test:create-block
Screen Shot 2021-03-23 at 10 38 25 AM

**Comment by gziolo** [comment]
> Folks might want to double check that package-lock.json is generated correctly, since the changes look a little large.

It's coming mostly from the Storybook upgrade. I guess it's inevitable with the number of changes they introduce 😞

**Comment by stokesman** [comment]
This seems to have broken `npm run dev` on my end. It completes the initial build without complaint but a moment after all the normal output it prints: `wp-scripts start exited with code 0`. The `npm run dev` job is still running and the file watching is too (updated files are noted in terminal) but the build does not run again.

When I encountered this I'd just merged trunk and noticing the package(-lock).json changes, I ran `npm install`. After rebuilds wouldn't work, I tried `npm run distclean` followed by another `npm install` but it didn't change. It started working again after reverting these changes and running `npm install`.

I'm using macOS 10.12.6. It doesn't seem like the changes here should have caused an issue. Maybe there's something else I could try besides the revert. Or a way I could get a better idea of what caused wp-scripts to exit. Any ideas would be much appreciated.

**Comment by noisysocks** [comment]
Yes something here breaks file watching in `npm run dev`. You can verify this quickly by running `npx wp-scripts start`. The process exits after building files once, but it should remain running until you manually terminate it using <kbd>Ctrl</kbd>+<kbd>C</kbd>. Reverting b90e38c and running `npm install` fixes the issue.

**Comment by gziolo** [comment]
Another attempt in #30347.

**Pull request merged**